File: rs/pushtofunction/src/fs/boot/foo_sfl_boot.rs
use crate::fs::{foo_sfl_c, FooSflCfg, FooSflDeps, FooSflT};
use common::config::AppCfgInfo;
use common::fs_data::FooSflCfgInfo;
use common::fwk::RefreshMode;
use std::sync::Arc;

use super::bar_bf_boot;

fn foo_sfl_cfg_adapter(app_cfg: &AppCfgInfo) -> FooSflCfgInfo {
    FooSflCfgInfo {
        a: app_cfg.x.clone(),
        b: app_cfg.y,
    }
}

pub fn foo_sfl_boot(app_cfg: fn() -> Arc<AppCfgInfo>, refresh_mode: RefreshMode) -> FooSflT {
    let cfg =
        FooSflCfg::new_boxed_with_cfg_adapter(app_cfg, foo_sfl_cfg_adapter, refresh_mode.clone());
    let deps = FooSflDeps {
        bar_bf: bar_bf_boot(app_cfg, refresh_mode),
    };
    foo_sfl_c(cfg, deps)
}

File: src/main/kotlin/pushtofunction/fs/boot/FooSflBoot.kt
/*
 * Copyright © 2022 Paulo Villela. All rights reserved.
 * Use of this source code is governed by the MIT license
 * that can be found in the LICENSE file.
 */

package pushtofunction.fs.boot

import pushtofunction.fs.FooSflCfgInfo
import pushtofunction.fs.FooSflCfgSrc
import pushtofunction.fs.FooSflT
import pushtofunction.fs.fooSflC
import pushtofunction.config.AppCfgInfo
import pushtovar.fwk.liftToNullary

private fun fooSflCfgAdapter0(appCfg: AppCfgInfo): FooSflCfgInfo {
	return FooSflCfgInfo(
		x = appCfg.x,
	)
}

var fooSflCfgAdapter = liftToNullary(::fooSflCfgAdapter0)

fun fooSflBoot(appCfg: (() -> AppCfgInfo)?): FooSflT {
	val fooSflCfgSrc = FooSflCfgSrc(
		get = fooSflCfgAdapter(appCfg),
		bar = barBfBoot(appCfg)
	)
	return fooSflC(fooSflCfgSrc)
}
